core/raftlease/store.go

```diff
@@ -0,0 +1,261 @@
+	FSM           ReadonlyFSM
+	Hub           *pubsub.StructuredHub
+	Target        NotifyTarget
+	RequestTopic  string
+	ResponseTopic func(requestID uint64) string
+
+	Clock          clock.Clock
+	ForwardTimeout time.Duration
+}
+
+// NewStore returns a core/lease.Store that manages leases in Raft.
+func NewStore(config StoreConfig) *Store {
+	return &Store{
+		fsm:      config.FSM,
+		hub:      config.Hub,
+		config:   config,
+		prevTime: config.FSM.GlobalTime(),
+	}
+}
+
+// Store manages a raft FSM and forwards writes through a pubsub hub.
+type Store struct {
+	fsm       ReadonlyFSM
+	hub       *pubsub.StructuredHub
+	requestID uint64
+	config    StoreConfig
+
+	prevTimeMu sync.Mutex
+	prevTime   time.Time
+}
+
+// ClaimLease is part of lease.Store.
+func (s *Store) ClaimLease(key lease.Key, req lease.Request) error {
+	err := s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationClaim,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+		Holder:    req.Holder,
+		Duration:  req.Duration,
+	})
+	if err == nil {
+		s.config.Target.Claimed(key, req.Holder)
+	}
+	return errors.Trace(err)
+}
+
+// ExtendLease is part of lease.Store.
+func (s *Store) ExtendLease(key lease.Key, req lease.Request) error {
+	return errors.Trace(s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationExtend,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+		Holder:    req.Holder,
+		Duration:  req.Duration,
+	}))
+}
+
+// ExpireLease is part of lease.Store.
+func (s *Store) ExpireLease(key lease.Key) error {
+	err := s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationExpire,
+		Namespace: key.Namespace,
+		ModelUUID: key.ModelUUID,
+		Lease:     key.Lease,
+	})
+	if err == nil {
+		s.config.Target.Expired(key)
+	}
+	return errors.Trace(err)
+}
+
+// Leases is part of lease.Store.
+func (s *Store) Leases() map[lease.Key]lease.Info {
+	leaseMap := s.fsm.Leases(s.config.Clock.Now())
+	result := make(map[lease.Key]lease.Info, len(leaseMap))
+	// Add trapdoors into the information from the FSM.
+	for k, v := range leaseMap {
+		v.Trapdoor = s.config.Target.Trapdoor(k, v.Holder)
+		result[k] = v
+	}
+	return result
+}
+
+// Refresh is part of lease.Store.
+func (s *Store) Refresh() error {
+	return nil
+}
+
+// Advance is part of globalclock.Updater.
+func (s *Store) Advance(duration time.Duration) error {
+	s.prevTimeMu.Lock()
+	defer s.prevTimeMu.Unlock()
+	newTime := s.prevTime.Add(duration)
+	err := s.forwardCommand(&Command{
+		Version:   CommandVersion,
+		Operation: OperationSetTime,
+		OldTime:   s.prevTime,
+		NewTime:   newTime,
+	})
+	if err == globalclock.ErrConcurrentUpdate {
+		// Someone else updated before us - get the new time.
+		s.prevTime = s.fsm.GlobalTime()
+	} else if err == nil {
+		s.prevTime = newTime
+	}
+	return errors.Trace(err)
+}
+
+func (s *Store) forwardCommand(command *Command) error {
```

Comment:
Maybe forwardCommandReturningResponseError?
Or something?
forwardCommand implies that the cmd is sent on and that's it, and any error returned is the result of that on-sending failing. Whereas we are sending on the command and waiting for a response back and if all that happens returning that error. Did we need to separate response error from marshall or send error? I guess not since we check for the error type and can deduce from that if it's a semantic lease related error

Reply:
You're right - forward is the wrong implication here. How about `runOnLeader`?

I don't think there's any point in splitting out a marshalling/transport error return from lease-specific errors here, because code that directly calls `runOnLeader` doesn't distinguish between those errors - the lease manager does but that's a layer away.